### PR TITLE
Custom pages

### PR DESCRIPTION
**Pull request by thynctank**
This adds SheetJS (xlsx) dependency for reports. May want to move into higher level and switch to package.json rather than committing locally, but it works for now.

### CODE REVIEW

File: custom-pages/gadgets/reports/reports.js [hunk 1]
```diff
@@ -1,13 +1,192 @@
...
+                let query = buildQuery(reportType)
+                if (query) {
+                    queryNodesThen(query, function () {
+                        workbook.SheetNames.push(reportType)
+                        workbook.Sheets[reportType] = buildWorksheet({
+                            reportType: reportType,
+                            nodes: this.asArray()
+                        })
+
+                        if (4 === workbook.SheetNames.length) {
```
Comment (xolotlatoani) [inline custom-pages/gadgets/reports/reports.js]: You can check against the size of `ReportTypes`

Comment (xolotlatoani) [reply custom-pages/gadgets/reports/reports.js]: or against the index that you are on

File: custom-pages/gadgets/reports/reports.js [hunk 2]
```diff
@@ -1,13 +1,192 @@
...
+            let rec = {}
+
+            //copy common fields
+            topLevelFieldsToCopy.forEach(function (fieldName) {
+                rec[fieldName] = record[fieldName]
+            })
+
+            rec._type = record.getTypeQName()
+            let meta = record.getSystemMetadata()
+            rec["_system.created_on.timestamp"] = meta.created_on.timestamp
```
Comment (xolotlatoani) [inline custom-pages/gadgets/reports/reports.js]: FYI,

you can also call meta.getCreatedOn()

File: custom-pages/gadgets/reports/reports.js [hunk 3]
```diff
@@ -1,13 +1,192 @@
...
+            rec["_system.created_by"] = meta.created_by
```
Comment (xolotlatoani) [inline custom-pages/gadgets/reports/reports.js]: same,

meta.getCreatedBy()

File: custom-pages/gadgets/reports/reports.js [hunk 4]
```diff
@@ -1,13 +1,192 @@
...
+            }
+
+            //copy over priceSkuList[0].price[ 0-2 ].priceType and priceValue
+            //copy over priceSkuList[0].sku[0].sol, soli, eol, and eoli
+            if (PRICESHEET === reportType) {
+                if (record.priceSkuList && record.priceSkuList[0]) {
+                    record.priceSkuList[0].price.forEach(function (price, index) {
+                        rec[`priceSkuList[0].price[${index}].priceType`] = price.priceType
+                        rec[`priceSkuList[0].price[${index}].priceValue`] = price.priceType
+                        rec[`priceSkuList[0].price[${index}].priceValue`] = price.priceType
```
Comment (xolotlatoani) [inline custom-pages/gadgets/reports/reports.js]: You're writing to the same property here

File: custom-pages/gadgets/reports/reports.js [hunk 5]
```diff
@@ -1,13 +1,192 @@
...
+                rec.skuId = record.skuId
+
+                if (record.color && record.color[0]) {
+                    rec.hexValue = record.color[0].hexValue
+                    rec.displayName = record.color[0].displayName
+                }
+            }
+
+            //copy over priceSkuList[0].price[ 0-2 ].priceType and priceValue
+            //copy over priceSkuList[0].sku[0].sol, soli, eol, and eoli
```
Comment (xolotlatoani) [inline custom-pages/gadgets/reports/reports.js]: I do not see these properties copied

**Review by xolotlatoani — CHANGES_REQUESTED** [review]

**Review by bengray — APPROVED** [review]

File: custom-pages/gadgets/reports/reports.js (hunk 1 above)
Comment (thynctank) [reply custom-pages/gadgets/reports/reports.js]: They're not guaranteed to finish in order requested so index is not usable. `ReportTypes` is a fixed length as it's just an array of consts

Comment (thynctank) [reply custom-pages/gadgets/reports/reports.js]: Have switched to using `ReportTypes.length`

File: custom-pages/gadgets/reports/reports.js (hunk 2 above)
Comment (thynctank) [reply custom-pages/gadgets/reports/reports.js]: Have switched all of the meta methods now.

File: custom-pages/gadgets/reports/reports.js (hunk 3 above)
Comment (thynctank) [reply custom-pages/gadgets/reports/reports.js]: See above

File: custom-pages/gadgets/reports/reports.js (hunk 4 above)
Comment (thynctank) [reply custom-pages/gadgets/reports/reports.js]: Ugh. Originally planned on adding other fields but forgot and I don't think they're needed anyway. Have removed

File: custom-pages/gadgets/reports/reports.js (hunk 5 above)
Comment (thynctank) [reply custom-pages/gadgets/reports/reports.js]: Yep. Originally had these in the CSV but have removed the comment as they are not used in xlsx for now

**Review by xolotlatoani — APPROVED** [review]

**Pull request merged**